### PR TITLE
Introduce fast path for buffered ReadableStream

### LINKED ISSUE #13678: Memory leak if accessing body of fetch() response before consuming the stream
### What version of Bun is running?

1.1.26+0a37423ba

### What platform is your computer?

Linux 6.6.16-linuxkit aarch64

### What steps can reproduce the bug?

Access 'body' of fetch() response before consuming the stream causes memory leak.

Code snippet: https://github.com/mshameti/bun-memory-accumulation-formdata-fetch-docker/blob/main/sender/script.js#L48-L72

The focus of this issue is LEAK 2. 

NOTE: This is the same repo used to report https://github.com/oven-sh/bun/issues/13657, but LEAK 2 is a different but closely related to LEAK 1. 

```
git clone https://github.com/mshameti/bun-memory-accumulation-formdata-fetch-docker
cd bun-memory-accumulation-formdata-fetch-docker
LEAK=2 docker-compose up --build
```

### What is the expected behavior?

No memory leak.

### What do you see instead?

Memory accumulation until container crashes.

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by Jarred-Sumner**
### What does this PR do?

Fixes #13678

This introduces a fast path for consuming streaming request bodies where previously the developer accessed the ReadableStream. This also fixes a memory leak when doing that.

Previously, the following code would be unnecessarily slow:
```js
Bun.serve({
  async fetch(req) {
    if (req.body) { 
       await req.json();
    }

    // previously, the fast path only worked if you didn't access req.body.
  }
});
```

This would cause us to:
1) Buffer the contents of the stream in JavaScript (slow)
2) High water mark of around 16 KB because we didn't implement the drain function for it.  

TODO:
- [x] Fix FormData issue
- [x] Fix other test failures
- [x] benchmark once CI finishes

### How did you verify your code works?

There are some new tests, and also I modified some existing tests to also cover this case. 

There's a lot of room for mistakes here. So let's be very confident this works properly when we merge it.

### CODE REVIEW

**Comment by robobun** [comment]
@Jarred-Sumner, your commit 84d07716fc04d48cef91d795900f2ad47079bc19 is building: [`#2672`](https://buildkite.com/bun/bun/builds/2672)

File: test/js/bun/http/bun-serve-static.test.ts [hunk 1]
```diff
@@ -143,7 +145,7 @@ describe("static", () => {
         Bun.gc(true);
 
         const rss = (process.memoryUsage.rss() / 1024 / 1024) | 0;
-        expect(rss).toBeLessThan(baseline * 4);
```
Comment (Jarred-Sumner) [inline test/js/bun/http/bun-serve-static.test.ts:145]: This test was not failing when it should've been failing. The test itself shouldn't use a ton of memory so I think a hardcoded value is fine here.

File: src/codegen/buildTypeFlag.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
+const buildTypeFlag = process.argv.find(argv => {
```
Comment (Jarred-Sumner) [inline src/codegen/buildTypeFlag.ts:1]: didn't mean to commit this

File: src/bun.js/webcore/response.zig [hunk 3]
```diff
@@ -1033,9 +1033,18 @@ pub const Fetch = struct {
...
+                        this.scheduled_response_buffer = .{
+                            .allocator = bun.default_allocator,
+                            .list = .{
+                                .items = &.{},
+                                .capacity = 0,
+                            },
+                        };
                         readable.ptr.Bytes.onData(
                             .{
-                                .temporary_and_done = bun.ByteList.initConst(chunk),
```
Comment (Jarred-Sumner) [inline src/bun.js/webcore/response.zig:1038]: subtle change

File: src/bun.js/bindings/Uint8Array.cpp [hunk 4]
```diff
@@ -25,4 +22,23 @@ extern "C" JSC::EncodedJSValue JSUint8Array__fromDefaultAllocator(JSC::JSGlobalO
...
     return JSC::JSValue::encode(uint8Array);
 }
+
+extern "C" JSC::EncodedJSValue JSArrayBuffer__fromDefaultAllocator(JSC::JSGlobalObject* lexicalGlobalObject, uint8_t* ptr, size_t length)
+{
+
+    JSC::JSArrayBuffer* arrayBuffer;
+
+    if (LIKELY(length > 0)) {
+        RefPtr<ArrayBuffer> buffer = ArrayBuffer::createFromBytes({ ptr, length }, createSharedTask<void(void*)>([](void* p) {
```
Comment (cirospaciari) [inline src/bun.js/bindings/Uint8Array.cpp:32]: should we handle exceptions here?

Comment (Jarred-Sumner) [reply src/bun.js/bindings/Uint8Array.cpp:32]: this line won't throw, it's a C++ ref counted ptr

File: src/codegen/buildTypeFlag.ts (hunk 2 above)
Comment (cirospaciari) [reply src/codegen/buildTypeFlag.ts:1]: can this be removed so do not cause confusion in the future?

**Review by cirospaciari — APPROVED** [review]
LGTM just the uneeded file that I think we should remove so we not associate with this PR in the future

**Review by cirospaciari — CHANGES_REQUESTED** [review]
Failing leak memory test.

**Comment by mshameti** [comment]
thank you for your hard work on this @Jarred-Sumner! the incredibly low turnaround time of issue to fix and enhancement is understated

**Pull request merged**